src/pcre2_compile.c

```diff
@@ -4301,9 +4314,17 @@ while (ptr < ptrend)
                 errorcode = ERR15;
                 goto FAILED;
                 }
-              *parsed_pattern++ = meta ? META_SCS_NEXT_NUMBER : META_SCS_NUMBER;
-              offset = (PCRE2_SIZE)(ptr - cb->start_pattern - 2);
-              PUTOFFSET(offset, parsed_pattern);
+
+              if (meta >= 2 && (next_offset - offset) <= 0xffff)
```

Comment:
How can you have `next_offset - offset > 0xffff`? Would that be if you have `(0000000000...0000001)` as the number in read_number, or does `(      'name'     )` allow skipping whitespace? Having >65000 characters of junk inside the scan_substring syntax seems perverse. You _could_ just reject `next_offset - offset > 0xffff` as invalid syntax, rather than having the non-RO versions of the META codes just to support this obscure case.

Reply:
There is a bit more complex problem here. The `compile_branch` function is recursive. Hence the initial offset of the initial meta code only known by the caller. The callee always use a non-RO variant first to initialize the offset, then use the RO variants.

Avoiding this behavior is difficult:
1) Doing it in the caller level: a loop could process the continuation meta codes. This would require a loop, which duplicates a lot of logic, or a sub-function.
2) Doing it in the callee level: the first opcode should go back in the meta code to find the starting offset. Looks a bit dangerous/fragile.
3) The meta code has an optional offset: the `length[meta]` structure has a fixed size for each element. This is also difficult to change.

I don't want to add space/comment support to the list at this point, because conditional blocks do not support them as well. This can be changed though.

I don't worry too much about the number of meta codes. We should never run out of them.

Anyway, if you have a better idea, let me know.

Reply:
I see, I see. Hmmm. I mainly find it a bit distasteful to see things handled in a non-uniform way.

We already have META_SCS_NAME and META_SCS_NEXT_NAME to handle the first and subsequent conditions separately; now you're creating META_SCS_NAME, META_SCS_N_NAME_RO, and META_SCS_N_NAME to handle the first three conditions differently!

How about this:

META_SCS simply starts the assertion. Contains no record of name or number. But does the recursive call into compile_regex(), which will consume the following META codes.

META_OFFSET always follows META_SCS, and will be consumed to set the offset inside the nested compile_regex().

Finally META_SCS_NUMBER and META_SCS_NAME search for the name/number and emit the appropriate OP_CREF/DNCREF. Both of these would embed a relative offset. If the offset is >0xffff then emit a parse error in the pre-parser.

The benefit would be: don't embed the first name/number into the SCS start, but instead handle the first one in the same way as the following ones. Also, we wouldn't then need more than one version of the anything.

Reply:
Ok, this would be a bigger rework then. I very much like this idea though. We have more characters here than `(?(` conditional blocks, since `(*scs:(` is at least 7 characters long. START(1), OFFSET(3), NUMBER/NAME(2) is still just 5.

I would still not worry about the > 0xffff case, since `META_OFFSET` could be repeated if needed.